README.md
# Change Log 1.9.0
 - Motor Control redesign
    - Added Motors Off 
    - Added an Extruder Page to control the extruder controls seperately
    - Added ability to heat the extruder from the extruder controls
 - Preheat Wizard
    - Added a feature to add custom Preheat settings
    - Added to the Fine Tune ZOffset Wizard, Change/Load Filament Wizard, Slicer Wizard. You can pick a custom heat for these wizards now
    - Added a Feature to edit/delete any preset
 - Z-Offset Wizard
    - Added in an extra 0.05mm
    - Z-Offset is now measured from the corner
 - Confirmation Screens
    - Added Confirmation screen to Reset EEPROM, Shutdown/Reboot, Reset Connection, and Delete File/Folder
 - Slicer Wizard
    - Added Brim and Raft Support
    - Added Supports from buildplate and everywhere
    - Added the ability to choose the temperature to print with
    - Added Fans on/off functionality
 - Added english.yaml file, which holds all the text for the program
 - Refactored Z-Offset Wizard to use template screens.
    - This change was to fit more language on the screen and auto adjust text as needed.
 - Added a kivy widget that auto adjusts image/text combos
 - Added a workflow to execute a function when backing out of a screen
 - Various text changes
 - Various Icon Changes
 - Bugfix
    - Fixed a bug that caused Anyone with dual extruders on the R2 to crash the screen during the Z-Offset Wizard
    - Fixed bug for disconnecting during an active print, the error message said "printing". It will not populate the correct error


# Change Log 1.8.0
 - Added a Web Cam Toggle Button

# Change Log 1.7.0
 - Print Tuning redesign
    - While printing print tuning will be shown, while not printing only fan controls will be shown
 - New Utilities Functions
    - Motors can be disengaged under Utilities >> Options >> Motors Off
    - You can now reset the connection to the printer at any time through the Utilities >> Option >> Connection tool 
    - Firmware Update has been moved to Options
 - Printer Status Redesign
    - New Non Blocking Error Messages
    - Temperature Controls and Motor Controls have been moved to the printer status screen
    - While Printing the screen will now show elapsed time and estimated time remaining
    - Added a progress bar
 - New Wizard Fine Tune Z Offset
 - New Wizard Bed Level Calibration (R2 Only)
 - fixed up and down buttons to have correctly sized elements
 - various language edits
 - Filament Load/Change will not mess with temperature while printing, it will also set the E-Steps back to where they were
 - Onboard sliced files will now be appended with meta data instead of saved directly with meta data


# Setup
If you want to install this repo:

### if you do not already have RoboLCD on the machine go ahead and clone it
```
git clone https://github.com/Robo3D/RoboLCD.git
```
### if you do have RoboLCD  on the machine pull the updates
```
# Make sure you are in the RoboLCD Folder
cd RoboLCD
git pull
```

### Switch to this branch
```
# Make sure you are in the RoboLCD Folder
cd RoboLCD
git checkout [branch name]
```

### install RoboLCD in develop mode
```
source ~/oprint/bin/activate
# Make sure you are in the RoboLCD Folder
cd ~/RoboLCD
python setup.py develop
```
### restart octoprint and pull up the log 
```
sudo service octoprint restart & tail -f ~/.octoprint/logs/octoprint.log
```





CHANGELOG.md
# RoboLCD Changelog

## 1.10.1 (2017-9-28)

### Bug Fixes

* Fixed: screen freezes when you try to print a file and your z offset is set outside range of 0 or -20. 

## 1.10.0 (2017-9-15)

### Improvements

* File system overhaul:
    * Supports gcode, stl, and hex files.
    * Supports local and usb mounted filesystems.
    * Sort by a-z, size, date, or file type.
    * Copy, move, or delete multiple files and folders at once.
    * Displays storage capacity used and total.
    * Meta data analysis performed when new file is selected for faster loading of file view.

* Reset connection screens now show connection status to printer's motor and temperature controls.

* Filament change will now push material through the nozzle before starting the retraction move. This mitigates filaments that get stuck and can't be pulled out easily.

* Added gcode command, 'F3000', to Raise Z option in motor controls so it would rise at an acceptable speed.


## Bug Fixes

* Fixed Network Title. It used to say Wizards

* Fixed the Error box starting up with a solid White image.

*  Fixed up and down buttons getting stuck on the File Screen

* Fixed Keyboard bug that would keep the keyboard up when an error occured.

*  Tool monitor will now show one decimal place. Graphical errors occured when showing two or more decimal places

## 1.9.1 (2017-7-21)

### Bug Fixes

* Fixed touch screen support for HDMI-powered C2 screens.

## 1.9.0 (2017-7-14)

### Improvements

* Motor control screen redesign.

* Custom preheat settings.

* Added 0.05mm increment to Z Offset Wizard.

* Z Offset Wizard measures z offset from corner for accuracy.

* Added confirmation screens to Reset Eeprom, shut down, reboot, reset connection, and delete files and folders

* Added more custom settings to Slicer Wizard: brim, build plate rafts, non-build plate rafts, temperature, and fans on.

### Bug Fixes

* Installing dual extruders do not crash the screen.

* Error message no longer displays "printing" when screen disconnects from motor contorls while printing. It will display real error message.

## 1.8.1 (2017-6-12)

### Bug Fixes

* Installation of RoboLCD does not ignore the new onboard webcam option created in 1.8.0.

## 1.8.0 (2017-6-9)

### Improvements

* New option: see the status of the onboard webcam: on or off; toggle its state (R2 only).

## 1.7.0 (2017-6-1)

### Improvements

* Print Tuning responsive redesign.

* Motors can be disengaged under Utilities>>Options>>Motors Offset.

* Reset connection to printer controls at any time through Utilities>>Options>>Connection.

* Firmware Update has been moved to Utilities>>Options.

* Printer Status screen responsive redesign.

* Support for Error Messages .

* Quick access to temperature controls and motor controls through printer status screen .

* Printing progress bar with elapsed time and remaining added to printer status screen.

* New Fine Tune Z offset wizard: calibrate your z offset to the hundredth of a mm. Can also be used to level your bed.

* New Bed Level Calibration wizard: guides you on how to level your R2's bed.(R2 only)

### Bug Fixes

* Various language edits.

* Filament Load/Change will not mess with temperature while printing, and will set the E-Steps back to normal.

* Onboard sliced files do not have missing meta data.



## 1.6.1 (2017-4-21)

### Bug Fixes

* Restarting octoprint no longer hangs when webcam is on.

* Hotspot mode retry button does not freeze screen.

* Removed connection delays between Octoprint and Mainboard.

* Adjust Z offset decimal styling error in wizard.

* Backing out of filament wizards does not freeze screen.

* QR code is bounded within blue boxes.


## 1.6.0 (2017-4-13)

### Improvements

* Update notification notifies users when OS updates are available

### Bug Fixes

* Screen styling is now compatible with R2 model

## 1.5.0 (2017-3-22)

### Improvements

* New file options available: add, move, traverse and delete folders; move and delete files.

* USB storage directory is mounted in the File's list. All new file options are available within the USB folder.

* Shortcut button added to Motor Control (top right): it will home all axes and move the bed up to the nozzle.

* Support for filament sensor


## 1.4.2 (2017-3-18)

### Bug Fixes

* R2 z offset wizard centers xy correctly

## 1.4.1 (2017-3-16)

### Bug Fixes

* R2 update screen no longer displays available C2 updates

## 1.4.0 (2017-3-15)

### Improvements

* Folder support  

* All usb files get dropped into a USB folder

* Scroll button can now be held down for faster scroll

* Slicing Wizard

* ZOffset is now a z-home offset instead of a z-probe offset. This generates more consistent and reliable z offsets.

### Bug Fixes

* Filament sensor no longer halts rendering of printer dashboard

* Fixed print error when printing from z offset warning popup

## 1.3.2 (2017-3-1)

### Bug Fixes

* Fixed issue where screen appears unresponsive when user pushes the update button: added a Popup that notifies user of background activity.  

##1.3.1 (2017-2-4)

###Bug Fixes

* Fixed issue with detecting R2 Bed

* Fixed issue where R2 bed disconnect freezes screen

* Added heated bed warning to avoid burns during z offset wizard

* Fans turn off after print is done


##1.3.0 (2017-2-3)

###Improvements

* Update Firmware from the Screen

* Display more gcode metadata: estimated time, z offset, infill, layer height, and layer count

* System Menu: Poweroff, Reboot and Restart Octoprint

* Options Menu: Unmount USB and Edit EEPROM settings

* Control temperature for dual extruders and heated bed (if applicable)

* Saves wifi password(s)

* Print tuning: fan speed, feed rate, and flow rate

* Z Offset wizard safeguard: turns off extruder before initiation to avoid melted beds

* Progress bar added for mounting usb files

###Bug Fixes

* Mounting USB can now handle 35+ gcode files

* Cannot update OS with no internet connection

* Mounting USB no longer causes loading delays


RoboLCD/lcd/files.py
from kivy.uix.label import Label
from kivy.uix.gridlayout import GridLayout
from kivy.uix.floatlayout import FloatLayout
from kivy.uix.boxlayout import BoxLayout
from kivy.uix.tabbedpanel import TabbedPanelHeader
from kivy.uix.button import Button
from kivy.uix.scrollview import ScrollView
from kivy.effects.scroll import ScrollEffect
from kivy.uix.screenmanager import Screen
from kivy.properties import ObjectProperty, StringProperty, BooleanProperty, NumericProperty
from kivy.clock import Clock
from kivy.uix.widget import Widget
from kivy.uix.popup import Popup
from pconsole import pconsole
import math
from datetime import datetime
from kivy.logger import Logger
from .. import roboprinter
import os
import shutil
import re
from scrollbox import ScrollBox, Scroll_Box_Even
import collections
from connection_popup import Zoffset_Warning_Popup, Error_Popup, USB_Progress_Popup
import subprocess
import threading
from session_saver import session_saver
import traceback
import octoprint.filemanager
from kivy.core.window import Window
from kivy.uix.vkeyboard import VKeyboard

from watchdog.observers import Observer
from watchdog.events import FileSystemEventHandler
from file_explorer import File_Explorer
from file_explorer import FileOptions
import re
from functools import partial


USB_DIR = '/home/pi/.octoprint/uploads/USB'
FILES_DIR = '/home/pi/.octoprint/uploads'

class FileButton(Button):
    def __init__(self, filename, date, path, is_usb = False, **kwargs):
        super(FileButton, self).__init__(**kwargs)

        self.filename = filename
        self.path = path
        self.date = date
        self.is_usb = is_usb
        self.long_press = False

        ##Format filename for display
        filename_no_ext = filename.replace('.gcode', '').replace('.gco', '')
        #Format spacing between filename and date
        if len(filename_no_ext) > 10:
            #filename_no_ext = filename_no_ext[:13] + '...' + filename_no_ext[-4:]
            self.ids.filename.text = filename_no_ext
            self.ids.date.text = date
        else:
            self.ids.filename.text = filename_no_ext
            self.ids.date.text = date

    def file_on_press(self):
        self.file_clock = Clock.schedule_once(self.file_clock_event, 1.0)
        pass

    def file_on_release(self):
        if not self.long_press:
            self.file_clock.cancel()
            roboprinter.robosm.generate_file_screen(filename=self.filename, file_path=self.path, is_usb=self.is_usb)
        else:
            self.long_press = False


    def file_clock_event(self, dt):
        # self.long_press = True
        # FileOptions(self.filename, self.path)
        pass # re enable for file options

    def open_file_options(self):
        FileOptions(self.filename, self.path)


class FilesTab(TabbedPanelHeader):
    """
    Represents the Files tab header and dynamic content
    """
    pass



class FilesContent(BoxLayout):
    """
    This class represents the properties and methods necessary to render the dynamic components of the FilesTab

    Files content is a Boxlayout made up of 2 boxes: files list (widget type == ScrollView) and scrolling buttons (widget type == Gridlayout)
    self.update() handles generating the Files content and it will generate only when the files list changes; such as when a user adds a file via the webapp or ios app
    """

    def __init__(self, **kwargs):
        super(FilesContent, self).__init__(**kwargs)
        Clock.schedule_interval(self.collect_meta_data, 0.1)

        #if octoprint has changed files then update them
        session_saver.save_variable('file_callback', self.call_to_update)
        session_saver.save_variable('usb_mounted', False)

        #schedule directory observer http://pythonhosted.org/watchdog/api.html#watchdog.events.FileSystemEventHandler
        self.dir_observe = Observer()
        self.callback = FileSystemEventHandler()
        self.callback.on_any_event = self.on_any_event
        self.dir_observe.schedule(self.callback, "/dev", recursive=True)
        self.dir_observe.start()


        self.update_files()


    # This seems like a roundabout way to call 'update_files' However if it is not called in this way 
    # Graphical issues become present.It is called in this way to make the request come from the 
    # thread kivy is on
    def call_to_update(self):
        Clock.schedule_once(self.call_to_update2, 0.01)
    def call_to_update2(self ,dt):
        self.update_files()

    def call_to_analyze(self,dt):

        roboprinter.printer_instance.start_analysis(files=session_saver.saved['current_files'])

    def update_files(self):
        files = File_Explorer('machinecode', FileButton, enable_editing=True)
        self.clear_widgets()
        self.add_widget(files)
        #helper function from the Meta Rader
        Clock.schedule_once(self.call_to_analyze, 0.01)

    #This function is a callback from an observer that is watching /dev for any device changes. Namely the USB being plugged in
    def on_any_event(self, event):
        usb_path = "/dev/sd"
        extern = re.match(usb_path, event.src_path)

        if extern != None and event.src_path.endswith('1'):

            if event.event_type == 'deleted':
                Logger.info("USB Removed " + event.src_path)
                session_saver.saved['usb_mounted'] = False
                self.has_usb_attached = False
                self.call_to_update()


            elif event.event_type == 'created':
                Logger.info("USB Plugged in " + event.src_path)
                session_saver.saved['usb_mounted'] = True
                session_saver.saved['usb_location'] = event.src_path
                self.has_usb_attached = True
                self.call_to_update()
        elif extern != None:
            Logger.info(event.src_path)

    #This function uses a shared funtion in the Meta Reader Plugin to collect information from a pipe
    #without disturbing the main thread
    def collect_meta_data(self, dt):
        roboprinter.printer_instance.collect_data()

class PrintFile(GridLayout):
    """
    This class encapsulates the dynamic properties that get rendered on the PrintFile and the methods that allow the user to start a print.
    """
    name = StringProperty('')
    backbutton_name = ObjectProperty(None)
    file_name = ObjectProperty(None)
    print_layer_height = ObjectProperty(None)
    print_layers = ObjectProperty(None)
    print_length = ObjectProperty(None)
    hours = NumericProperty(0)
    minutes = NumericProperty(0)
    seconds = NumericProperty(0)
    infill = ObjectProperty(None)
    file_path = ObjectProperty(None)
    status = StringProperty('--')
    subtract_amount = NumericProperty(30)
    current_z_offset = StringProperty('--')

    def __init__(self, **kwargs):
        super(PrintFile, self).__init__(**kwargs)

        self.status = self.is_ready_to_print()
        Clock.schedule_interval(self.update, .1)

        self.current_z_offset = str("{0:.1f}".format(float(pconsole.home_offset['Z'])))

        cura_meta = self.check_saved_data()
        self.print_layer_height = '--'
        self.print_layers = '--'
        self.infill = '--'
        self.hours = 0
        self.minutes = 0
        self.seconds = 0

        if cura_meta != False:
            if 'layer height' in cura_meta:
                self.print_layer_height = cura_meta['layer height']
            else:
                self.print_layer_height = "--"
            if 'layers' in cura_meta:
                self.print_layers = cura_meta['layers']
            else:
                layers = "--"
            if 'infill' in cura_meta:
                self.infill = cura_meta['infill']
            else:
                infill = "--"

            if 'time' in cura_meta:
                self.hours = int(cura_meta['time']['hours'])
                self.minutes = int(cura_meta['time']['minutes'])
                self.seconds = int(cura_meta['time']['seconds'])
            else:
                self.hours = 0
                self.minutes = 0
                self.seconds = 0

        else:
            self.print_layer_height = '--'
            self.print_layers = '--'
            self.infill = '--'
            self.hours = 0
            self.minutes = 0
            self.seconds = 0

    #This function will check the filename against saved data on the machine and return saved meta data
    def check_saved_data(self):
        self.octo_meta = roboprinter.printer_instance._file_manager
        saved_data = self.octo_meta.get_metadata(octoprint.filemanager.FileDestinations.LOCAL , self.file_path)


        if 'robo_data' in saved_data:
            return saved_data['robo_data']
        else:
            return False


    def update(self, dt):
        if roboprinter.printer_instance._printer.is_paused():
            self.status = 'PRINTER IS BUSY'
        else:
            self.status = self.is_ready_to_print()
        #toggle between button states
        if self.status == 'PRINTER IS BUSY' and self.ids.start.background_normal == "Icons/green_button_style.png":
            self.ids.start.background_normal = "Icons/red_button_style.png"
            self.ids.start.button_text = '[size=30]' + roboprinter.lang.pack['Files']['File_Status']['Busy'] + '[/size]'
            self.ids.start.image_icon = 'Icons/Manual_Control/printer_button_icon.png'
         
        elif self.status == "READY TO PRINT" and self.ids.start.background_normal == "Icons/red_button_style.png":
            self.ids.start.background_normal = "Icons/green_button_style.png"
            self.ids.start.button_text = '[size=30]' + roboprinter.lang.pack['Files']['File_Status']['Start'] + '[/size]'
            self.ids.start.image_icon = 'Icons/Manual_Control/start_button_icon.png'
            

    def start_print(self, *args):
        #Throw a popup to display the ZOffset if the ZOffset is -10 or more
        try:
            if self.status == "READY TO PRINT":
                _offset = float(pconsole.home_offset['Z'])

                if _offset <= -20.0 or _offset >= 0.0:
                    zoff = Zoffset_Warning_Popup(self)
                    zoff.open()
                else:
                    """Starts print but cannot start a print when the printer is busy"""
                    Logger.info(self.file_path)
                    self.force_start_print()
        except Exception as e:
            #raise error
            error = Error_Popup(roboprinter.lang.pack['Files']['File_Error']['Title'],roboprinter.lang.pack['Files']['File_Error']['Body'],callback=partial(roboprinter.robosm.go_back_to_main, tab='printer_status_tab'))
            error.open()
            Logger.info("Start Print Error")
            Logger.info("!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!! "+ str(e))
            traceback.print_exc()

    def force_start_print(self, *args):
        """Starts print but cannot start a print when the printer is busy"""
        try:
            path_on_disk = roboprinter.printer_instance._file_manager.path_on_disk(octoprint.filemanager.FileDestinations.LOCAL, self.file_path)
            roboprinter.printer_instance._printer.select_file(path=path_on_disk, sd=False, printAfterSelect=True)
        except Exception as e:
            #raise error
            error = Error_Popup(roboprinter.lang.pack['Files']['File_Error']['Title'],roboprinter.lang.pack['Files']['File_Error']['Body'],callback=partial(roboprinter.robosm.go_back_to_main, tab='printer_status_tab'))
            error.open()
            Logger.info("Force Start Print Error")
            Logger.info("!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!! "+ str(e))
            traceback.print_exc()


    def is_ready_to_print(self):
        """ whether the printer is currently operational and ready for a new print job"""
        is_ready = roboprinter.printer_instance._printer.is_ready()
        printing = roboprinter.printer_instance._printer.is_printing()
        if is_ready and not printing:
            return 'READY TO PRINT'
        else:
            return 'PRINTER IS BUSY'

class PrintUSB(PrintFile):
    """
        This class encapsulates the dynamic properties that get rendered on the PrintUSB and the methods that allow the user to start a print from usb or save the file to local.
    """
    def __init__(self, **kwargs):
        super(PrintUSB, self).__init__(**kwargs)
        self.progress_pop =  USB_Progress_Popup("Saving File", 1)
        pass


    def save_file_to_local(self, *args):
        self.progress_pop.show()

        Clock.schedule_once(self.attempt_to_save, 0.01)



    def attempt_to_save(self, dt):
        try:
            copy_path = FILES_DIR + '/' + self.file_name
            real_path = roboprinter.printer_instance._file_manager.path_on_disk('local', self.file_path)

            #shutil.copy2(real_path, copy_path)
            Logger.info("Started the Copy src: " + real_path + " cp to dst: " + copy_path )
            copied = self.copy_file(real_path, copy_path, progress_callback=self.progress_update)
            if not copied:
                self.progress_pop.hide()
                ep = Error_Popup(roboprinter.lang.pack['Files']['File_Error']['Title'],roboprinter.lang.pack['Files']['File_Error']['Body'],callback=partial(roboprinter.robosm.go_back_to_main, tab='printer_status_tab'))
                ep.show()
                Logger.info("attempt to save Error")


        except Exception as e:
            #raise error
            self.progress_pop.hide()
            ep = Error_Popup(roboprinter.lang.pack['Files']['File_Error']['Title'],roboprinter.lang.pack['Files']['File_Error']['Body'],callback=partial(roboprinter.robosm.go_back_to_main, tab='printer_status_tab'))
            ep.show()
            Logger.info("!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!! "+ str(e))
            traceback.print_exc()


    def copy_file(self, fsrc, fdst, progress_callback=None, complete_callback = None, length=16*1024, **kwargs):

        self.copied = 0
        self.file_size = 0
        self.length = length
        self.p_callback = progress_callback
        self.c_callback = complete_callback
        if not os.path.isfile(fsrc):
            Logger.info("Will not copy")
            return False

        else:
            self.file_size = float(os.path.getsize(fsrc))
        #make the new file
        self.src_obj = open(fsrc, 'rb')
        self.dst_obj = open(fdst, 'wb')
        #Do the copy as fast as possible without blocking the UI thread
        Clock.schedule_interval(self.copy_object, 0)
        return True

    #doing it this way with a clock object does not block the UI
    def copy_object(self, dt):
        #grab part of the file
        buf = self.src_obj.read(self.length)
        #if there isn't anything to read then close the files and return
        if not buf:
            self.src_obj.close()
            self.dst_obj.close()
            if self.c_callback != None:
                self.c_callback()
            return False
        #Write the buffer to the new file
        self.dst_obj.write(buf)
        #update how much of the file has been copied
        self.copied += len(buf)

        #report progress
        if self.p_callback != None:
            progress = float(self.copied/self.file_size)
            self.p_callback(progress)

    def progress_update(self, progress):
        self.progress_pop.update_progress(progress)
        #Logger.info(str(progress))

        if progress == 1.0:
            self.progress_pop.hide()
            ep = Error_Popup(roboprinter.lang.pack['Files']['File_Saved']['Title'] , roboprinter.lang.pack['Files']['File_Saved']['Body'],callback=partial(roboprinter.robosm.go_back_to_main, tab='printer_status_tab'))
            ep.show()
            if 'file_callback' in session_saver.saved:
                session_saver.saved['file_callback']()
